src/main/scala/example/Validator.scala

```diff
@@ -0,0 +1,91 @@
+package example
+
+/**
+  * Implement validator typeclass that should validate arbitrary value [T].
+  * @tparam T the type of the value to be validated.
+  */
+trait Validator[T] {
+  /**
+    * Validates the value.
+    * @param value value to be validated.
+    * @return Right(value) in case the value is valid, Left(message) on invalid value
+    */
+  def validate(value: T): Either[String, T]
+
+  /**
+    * And combinator.
+    * @param other validator to be combined with 'and' with this validator.
+    * @return the Right(value) only in case this validator and <code>other</code> validator returns valid value,
+    *         otherwise Left with error messages from both validators.
+    */
+  def and(other: Validator[T]): Validator[T] = (value: T) => {
+    val first = this.validate(value)
+    val second = other.validate(value)
+    Either.cond(first.isRight && second.isRight, value, first.left.getOrElse("") + second.left.getOrElse(""))
+  }
+
+  /**
+    * Or combinator.
+    * @param other validator to be combined with 'or' with this validator.
+    * @return the Right(value) only in case either this validator or <code>other</code> validator returns valid value,
+    *         otherwise Left with error messages from the failed validator or from both if both failed.
+    */
+  def or(other: Validator[T]): Validator[T] = (value: T) => {
+    val first = this.validate(value)
+    val second = other.validate(value)
+    Either.cond(first.isRight || second.isRight, value, first.left.getOrElse("") + second.left.getOrElse(""))
+  }
+}
+
+
+object Validator {
+  val positiveInt : Validator[Int] = (t: Int) => Either.cond(t > 0, t, "Value should be positive. ")
+
+  def lessThan(n: Int): Validator[Int] = (t: Int) => Either.cond(t < n, t, "Value should be less than given. ")
+
+  val nonEmpty : Validator[String] = (t: String) => Either.cond(t.nonEmpty, t, "Value should not be empty. ")
+
+  val isPersonValid = new Validator[Person] {
+    // Returns valid only when the name is not empty and age is in range [1-99].
+    override def validate(value: Person): Either[String, Person] = Either.cond(value.name.nonEmpty && (1 to 99).contains(value.age),
+        value, "Person is invalid - either name is empty, either age is not in range [1-99]. ")
+  }
+
+  implicit class IntImprovements(val n: Int) {
+    def validate(implicit t:Validator[Int] = Validator.positiveInt): Either[String, Int] = t.validate(n)
```

Comment:
There is no need to add a default value for implicit parameter. Compiler will find `Validator` instances if they are declared as `implicit`